src/crimson/os/seastore/lba_mapping.h

```diff
@@ -10,125 +10,198 @@
+    return physical_cursor->val.pladdr.get_paddr();
+  }
+
+  uint32_t get_checksum() const {
+    assert(!is_null());
+    return physical_cursor->val.checksum;
   }
 
-  virtual void maybe_fix_pos() = 0;
-  virtual bool parent_modified() const = 0;
-  virtual uint32_t get_checksum() const = 0;
+  laddr_t get_key() const {
+    assert(!is_null());
+    if (is_indirect()) {
+      return indirect_cursor->key;
+    } else {
+      return physical_cursor->key;
+    }
+  }
 
   // An lba pin may be indirect, see comments in lba_manager/btree/btree_lba_manager.h
-  virtual bool is_indirect() const = 0;
-  virtual laddr_t get_intermediate_key() const = 0;
-  virtual laddr_t get_intermediate_base() const = 0;
-  virtual extent_len_t get_intermediate_length() const = 0;
-  // The start offset of the pin, must be 0 if the pin is not indirect
-  virtual extent_len_t get_intermediate_offset() const = 0;
+  laddr_t get_intermediate_key() const {
+    assert(!is_null());
+    assert(is_indirect());
+    return indirect_cursor->val.pladdr.get_laddr();
+  }
+  laddr_t get_intermediate_base() const {
+    assert(!is_null());
+    assert(is_indirect());
+    return physical_cursor->key;
+  }
+  extent_len_t get_intermediate_length() const {
+    assert(!is_null());
+    assert(is_indirect());
+    return physical_cursor->val.len;
+  }
+  // The start offset of the indirect cursor related to physical cursor
+  extent_len_t get_intermediate_offset() const {
+    assert(!is_null());
+    assert(is_indirect());
+    return get_intermediate_base().get_byte_distance<
+      extent_len_t>(get_intermediate_key());
+  }
 
-  virtual get_child_ret_t<lba_manager::btree::LBALeafNode, LogicalChildNode>
-  get_logical_extent(Transaction &t) = 0;
+  get_child_ret_t<lba_manager::btree::LBALeafNode, LogicalChildNode>
+  get_logical_extent(Transaction &t);
 
-  void link_child(LogicalChildNode *c) {
-    ceph_assert(child_pos);
-    child_pos->link_child(c);
+  void link_child(LogicalChildNode &extent) {
+    using lba_manager::btree::LBALeafNode;
+    auto leaf = physical_cursor->parent->cast<LBALeafNode>();
+    assert(leaf->is_stable());
+    leaf->link_child(&extent, physical_cursor->pos);
   }
-  virtual LBAMappingRef refresh_with_pending_parent() = 0;
 
-  // For reserved mappings, the return values are
-  // undefined although it won't crash
-  virtual bool is_stable() const = 0;
-  virtual bool is_data_stable() const = 0;
-  virtual bool is_clone() const = 0;
-  bool is_zero_reserved() const {
-    return !get_val().is_real();
+  LBAMapping duplicate() const {
+    assert(!is_null());
+    auto dup_iter = [](const LBACursorRef &iter) -> LBACursorRef {
+      if (iter) {
+	return std::make_unique<LBACursor>(*iter);
+      } else {
+	return nullptr;
+      }
+    };
+    return LBAMapping(dup_iter(physical_cursor), dup_iter(indirect_cursor));
   }
 
-  LBAMappingRef duplicate() const;
+private:
+  // Only allow BtreeLBAManager use these two methods
+  // FIXME: remove them
```

Comment:
What do you mean by "remove them" here ?

Reply:
I see, there should be no difference between make_indirect and link_physical now.

Reply:
As explained in https://github.com/ceph/ceph/pull/62071#discussion_r1986761211 , the mapping is at a temporary state, LBAManager should use these two methods to complete the construction of LBAMapping.

If we could move the logic about remapping and cloning out of LBAManager, the LBAMapping can be constructed directly without `make_indirect()` and `link_direct()`.

Reply:
See https://github.com/ceph/ceph/pull/62071#discussion_r2020605859